diff -u -r -N qt-win-opensource-src-4.2.0-snapshot-20060921-orig/demos/dbus-viewer/dbus-viewer.pro demos/dbus-viewer/dbus-viewer.pro
--- tools/qdbus/tools/qdbuscpp2xml/qdbuscpp2xml.cpp	Thu Apr 12 13:22:27 CEST 2007
+++ tools/qdbus/tools/qdbuscpp2xml/qdbuscpp2xml.cpp	Thu Apr 12 13:22:27 CEST 2007
@@ -366,7 +366,7 @@
         else {
             // run moc on this file
             QProcess proc;
-            proc.start(QLibraryInfo::location(QLibraryInfo::BinariesPath) + QLatin1String("/moc"), QStringList() << QFile::decodeName(argv[i]));
+            proc.start(QLibraryInfo::location(QLibraryInfo::BinariesPath) + QLatin1String("/moc"), QStringList() << QFile::decodeName(argv[i]), QIODevice::ReadOnly | QIODevice::Text);
 
             if (!proc.waitForStarted()) {
                 fprintf(stderr, PROGRAMNAME ": could not execute moc! Aborting.\n");
diff -u -r -N qt-win-opensource-src-4.2.0-snapshot-20060921-orig/README-qdbus.txt README-qdbus.txt
--- qt-win-opensource-src-4.2.0-snapshot-20060921-orig/README-qdbus.txt	1970-01-01 01:00:00.000000000 +0100
+++ README-qdbus.txt	2006-09-21 16:32:40.000000000 +0200
@@ -0,0 +1,5 @@
+to activate qdbus support add the line
+QT_CONFIG += qdbus
+and the following in case of mingw compiler
+CONFIG += sse
+to .qmake.cache after configuring and run your make tool
