diff -u -r -N qt-win-opensource-src-4.2.0-snapshot-20060921-orig/demos/dbus-viewer/dbus-viewer.pro demos/dbus-viewer/dbus-viewer.pro
--- tools\qdbus\tools\qdbuscpp2xml\qdbuscpp2xml.cpp.orig	2006-10-20 18:27:34.000000000 +0200
+++ tools\qdbus\tools\qdbuscpp2xml\qdbuscpp2xml.cpp	2006-12-21 08:58:42.515625000 +0100
@@ -95,7 +95,7 @@
 QByteArray MocParser::readLine()
 {
     ++lineNumber;
-    return input->readLine();
+    return input->readLine().replace("\r\n", "\n");
 }
 
 void MocParser::loadIntData(uint *&data)
diff -u -r -N qt-win-opensource-src-4.2.0-snapshot-20060921-orig/mkspecs/features/qdbus.prf mkspecs/features/qdbus.prf
--- qt-win-opensource-src-4.2.0-snapshot-20060921-orig/mkspecs/features/qdbus.prf	2006-09-21 01:23:30.000000000 +0200
+++ mkspecs/features/qdbus.prf	2006-09-21 16:31:04.000000000 +0200
@@ -1,6 +1,10 @@
 
 qtAddLibrary(QtDBus)
 
+win32: !contains(QT, gui):contains(TEMPLATE, app) {
+	CONFIG += console
+}
+
 isEmpty(QMAKE_QDBUSXML2CPP) {
     win32:QMAKE_QDBUSXML2CPP = $$[QT_INSTALL_BINS]\qdbusxml2cpp.exe
     else:QMAKE_QDBUSXML2CPP = $$[QT_INSTALL_BINS]/qdbusxml2cpp
diff -u -r -N qt-win-opensource-src-4.2.0-snapshot-20060921-orig/README-qdbus.txt README-qdbus.txt
--- qt-win-opensource-src-4.2.0-snapshot-20060921-orig/README-qdbus.txt	1970-01-01 01:00:00.000000000 +0100
+++ README-qdbus.txt	2006-09-21 16:32:40.000000000 +0200
@@ -0,0 +1,5 @@
+to activate qdbus support add the line
+QT_CONFIG += qdbus
+and the following in case of mingw compiler
+CONFIG += sse
+to .qmake.cache after configuring and run your make tool
