Index: dbus/dbus-internals.h
===================================================================
--- dbus/dbus-internals.h       (revision 64)
+++ dbus/dbus-internals.h       (working copy)
@@ -281,7 +281,9 @@
 _DBUS_DECLARE_GLOBAL_LOCK (system_users);
 _DBUS_DECLARE_GLOBAL_LOCK (message_cache);
 _DBUS_DECLARE_GLOBAL_LOCK (shared_connections);
-#define _DBUS_N_GLOBAL_LOCKS (11)
+_DBUS_DECLARE_GLOBAL_LOCK (win32_fds);
+_DBUS_DECLARE_GLOBAL_LOCK (sid_atom_cache);
+#define _DBUS_N_GLOBAL_LOCKS (13)

 dbus_bool_t _dbus_threads_init_debug (void);
