Index: bus/main.c
===================================================================
--- bus/main.c	(Revision 375)
+++ bus/main.c	(Arbeitskopie)
@@ -234,6 +233,12 @@
   dbus_bool_t print_pid;
   int force_fork;
 
+#ifdef DBUS_WIN
+	extern int _dbus_init_working_dir(char *s);
+	if (!_dbus_init_working_dir(argv[0]))
+		return 1;
+#endif
+
   if (!_dbus_string_init (&config_file))
     return 1;
 
Index: bus/config-loader-libxml.c
===================================================================
--- bus/config-loader-libxml.c	(Revision 375)
+++ bus/config-loader-libxml.c	(Arbeitskopie)
@@ -177,13 +177,19 @@
       xmlSetGenericErrorFunc (NULL, xml_shut_up);
     }
 
+#ifdef DBUS_WIN
+	DBusString *_dbus_get_working_dir(void);
+	DBusString *dbusdir = _dbus_get_working_dir();
+  parser = bus_config_parser_new (dbusdir, is_toplevel, parent);
+#else
   if (!_dbus_string_get_dirname (file, &dirname))
     {
       _DBUS_SET_OOM (error);
       goto failed;
     }
+  parser = bus_config_parser_new (&dirname, is_toplevel, parent);
+#endif
   
-  parser = bus_config_parser_new (&dirname, is_toplevel, parent);
   if (parser == NULL)
     {
       _DBUS_SET_OOM (error);
