Index: dbus/dbus-internals.h
===================================================================
--- dbus/dbus-internals.h	(revision 481)
+++ dbus/dbus-internals.h	(working copy)
@@ -48,7 +48,7 @@
 
 #if defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 199901L)
 #define _DBUS_FUNCTION_NAME __func__
-#elif defined(__GNUC__)
+#elif defined(__GNUC__) || defined(_MSC_VER)
 #define _DBUS_FUNCTION_NAME __FUNCTION__
 #else
 #define _DBUS_FUNCTION_NAME "unknown function"
