Index: dbus/dbus-marshal-recursive.c
===================================================================
--- dbus/dbus-marshal-recursive.c	(revision 532)
+++ dbus/dbus-marshal-recursive.c	(working copy)
@@ -690,7 +690,7 @@
   base_reader_next
 };
 
-static const DBusTypeReaderClass const *
+static const DBusTypeReaderClass * const
 all_reader_classes[] = {
   &body_reader_class,
   &body_types_only_reader_class,
