Index: dbus/dbus-internals.h
===================================================================
--- dbus/dbus-internals.h       (revision 49)
+++ dbus/dbus-internals.h       (working copy)
@@ -81,9 +81,16 @@
                                ...) _DBUS_GNUC_PRINTF (1, 2);
 void _dbus_verbose_reset_real (void);

-#  define _dbus_verbose _dbus_verbose_real
+#  ifdef HAVE_GNUC_VARARGS
+#    define _dbus_verbose(format, rest...) _dbus_verbose_real("[%s %d] " format, __FUNCTION__,__LINE__, ## rest)
+#  else
+#    define _dbus_verbose _dbus_verbose_real
+#  endif // HAVE_GNUC_VARARGS
+
 #  define _dbus_verbose_reset _dbus_verbose_reset_real
-#else
+
+#else // DBUS_ENABLE_VERBOSE_MODE
+
 