Index: test/CVS/Entries
===================================================================
--- test/CVS/Entries	(revision 610)
+++ test/CVS/Entries	(working copy)
@@ -1,3 +1,3 @@
-/CMakeLists.txt/1.1/Sat Mar  3 14:32:17 2007//
+/CMakeLists.txt/1.2/Sat Apr 28 19:37:25 2007//
 D/data////
 D/name-test////
Index: test/name-test/CVS/Entries
===================================================================
--- test/name-test/CVS/Entries	(revision 610)
+++ test/name-test/CVS/Entries	(working copy)
@@ -1,2 +1,2 @@
-/CMakeLists.txt/1.1/Sat Mar  3 14:32:17 2007//
+/CMakeLists.txt/1.2/Sat Apr 28 19:37:25 2007//
 D
Index: test/name-test/CMakeLists.txt
===================================================================
--- test/name-test/CMakeLists.txt	(revision 610)
+++ test/name-test/CMakeLists.txt	(working copy)
@@ -9,17 +9,17 @@
 # there is also a test-names executable, don't know if this is the same
 add_executable(test-names1 ${NAMEtest-DIR}/test-names.c)
 target_link_libraries(test-names1 dbus-1)
-install_targets(/bin test-names1)
+#install_targets(/bin test-names1)
 ADD_TEST(test-names ${EXECUTABLE_OUTPUT_PATH}/test-names1)
 
 add_executable(test-pending-call-dispatch ${NAMEtest-DIR}/test-pending-call-dispatch.c)
 target_link_libraries(test-pending-call-dispatch dbus-1)
-install_targets(/bin test-pending-call-dispatch)
+#install_targets(/bin test-pending-call-dispatch)
 ADD_TEST(test-pending-call-dispatch ${EXECUTABLE_OUTPUT_PATH}/test-pending_call-dispatch)
 
 add_executable(test-thread-init ${NAMEtest-DIR}/test-threads-init.c)
 target_link_libraries(test-thread-init dbus-1)
-install_targets(/bin test-thread-init)
+#install_targets(/bin test-thread-init)
 ADD_TEST(test-names ${EXECUTABLE_OUTPUT_PATH}/test-thread-init)
 
 endif (DBUS_BUILD_TESTS)
Index: test/CMakeLists.txt
===================================================================
--- test/CMakeLists.txt	(revision 610)
+++ test/CMakeLists.txt	(working copy)
@@ -90,7 +90,7 @@
 
 add_executable(test-service ${test-service_SOURCES})
 target_link_libraries(test-service dbus-1)
-install_targets(/bin test-service)
+#install_targets(/bin test-service)
 
 # (rh) does not work at now
 ADD_TEST(test-service ${EXECUTABLE_OUTPUT_PATH}/test-service${EXT})
@@ -99,17 +99,17 @@
 
 add_executable(test-names ${test-names_SOURCES})
 target_link_libraries(test-names dbus-1)
-install_targets(/bin test-names)
+#install_targets(/bin test-names)
 ADD_TEST(test-names ${EXECUTABLE_OUTPUT_PATH}/test-names${EXT})
 
 add_executable(test-shell ${test-shell_SOURCES})
 target_link_libraries(test-shell dbus-1)
-install_targets(/bin test-shell)
+#install_targets(/bin test-shell)
 ADD_TEST(test-shell ${EXECUTABLE_OUTPUT_PATH}/test-shell${EXT})
 
 add_executable(test-shell-service ${test-shell_SOURCES})
 target_link_libraries(test-shell-service dbus-1)
-install_targets(/bin test-shell-service)
+#install_targets(/bin test-shell-service)
 ADD_TEST(test-shell-service ${EXECUTABLE_OUTPUT_PATH}/test-shell-service${EXT})
 
 add_executable(test-spawn ${test-spawn_SOURCES})
@@ -119,12 +119,12 @@
 
 add_executable(test-exit ${test-exit_SOURCES})
 target_link_libraries(test-exit dbus-1)
-install_targets(/bin test-exit)
+#install_targets(/bin test-exit)
 ADD_TEST(test-exit ${EXECUTABLE_OUTPUT_PATH}/test-exit${EXT})
 
 add_executable(test-segfault ${test-segfault_SOURCES})
 target_link_libraries(test-segfault dbus-1)
-install_targets(/bin test-segfault)
+#install_targets(/bin test-segfault)
 ADD_TEST(test-segfault ${EXECUTABLE_OUTPUT_PATH}/test-segfault${EXT})
 
 add_executable(test-sleep-forever ${test-sleep-forever_SOURCES})
@@ -203,7 +203,7 @@
 		MESSAGE("FROM: ${FILE}\nTO: ${TARGET}\n")
 	ENDIF (CONFIG_VERBOSE)
 	configure_file(${FILE} ${TARGET} )
-	install_files(/data/dbus-1/services FILES ${TARGET})
+	#install_files(/data/dbus-1/services FILES ${TARGET})
 ENDFOREACH(FILE)
 
 #
Index: tools/CVS/Entries
===================================================================
--- tools/CVS/Entries	(revision 610)
+++ tools/CVS/Entries	(working copy)
@@ -1,2 +1,2 @@
-/CMakeLists.txt/1.1/Sat Mar  3 14:32:17 2007//
+/CMakeLists.txt/1.2/Thu Apr 26 09:30:38 2007//
 D
Index: tools/CMakeLists.txt
===================================================================
--- tools/CMakeLists.txt	(revision 610)
+++ tools/CMakeLists.txt	(working copy)
@@ -54,16 +54,16 @@
 	../../tools/dbus-viewer.c
 )
 
-add_executable(dbus-send ${dbus_send_SOURCES})
-target_link_libraries(dbus-send dbus-1)
-install_targets(/bin dbus-send )
+add_executable(dbus-send${CMAKE_EXE_POSTFIX} ${dbus_send_SOURCES})
+target_link_libraries(dbus-send${CMAKE_EXE_POSTFIX} dbus-1)
+install_targets(/bin dbus-send${CMAKE_EXE_POSTFIX} )
 
 # glib required 
-#add_executable(dbus_launch ${dbus_launch_SOURCES})
+#add_executable(dbus_launch${CMAKE_EXE_POSTFIX} ${dbus_launch_SOURCES})
 
-add_executable(dbus-monitor ${dbus_monitor_SOURCES})
-target_link_libraries(dbus-monitor dbus-1)
-install_targets(/bin dbus-monitor )
+add_executable(dbus-monitor${CMAKE_EXE_POSTFIX} ${dbus_monitor_SOURCES})
+target_link_libraries(dbus-monitor${CMAKE_EXE_POSTFIX} dbus-1)
+install_targets(/bin dbus-monitor${CMAKE_EXE_POSTFIX} )
 
 #dbus_send_LDADD= $(top_builddir)/dbus/libdbus-1.la
 #dbus_monitor_LDADD= $(top_builddir)/glib/libdbus-glib-1.la
Index: modules/FindLibExpat.cmake
===================================================================
--- modules/FindLibExpat.cmake	(revision 610)
+++ modules/FindLibExpat.cmake	(working copy)
@@ -13,13 +13,23 @@
 
 else (LIBEXPAT_INCLUDE_DIR AND LIBEXPAT_LIBRARIES)
 
-    IF (NOT WIN32)
+    IF (WIN32)
+		file(TO_CMAKE_PATH "$ENV{PROGRAMFILES}" _progFiles)
+		find_FILE(LIBEXPAT_DIR expat Source/lib/expat.h
+   			PATHS
+   			"${_progFiles}"
+		)
+        if (LIBEXPAT_DIR)
+            set (_LIBEXPATIncDir  ${LIBEXPAT_DIR}/Source/lib)
+            set (_LIBEXPATLinkDir ${LIBEXPAT_DIR}/libs)
+        endif (LIBEXPAT_DIR)
+    ELSE (WIN32)
         # use pkg-config to get the directories and then use these values
         # in the FIND_PATH() and FIND_LIBRARY() calls
         INCLUDE(UsePkgConfig)
         PKGCONFIG(LIBEXPAT-2.0 _LIBEXPATIncDir _LIBEXPATLinkDir _LIBEXPATLinkFlags _LiIconvCflags)
         SET(LIBEXPAT_DEFINITIONS ${_LIBEXPATCflags})
-    ENDIF (NOT WIN32)
+    ENDIF (WIN32)
 
     FIND_PATH(LIBEXPAT_INCLUDE_DIR expat.h
       PATHS
Index: modules/CVS/Entries
===================================================================
--- modules/CVS/Entries	(revision 610)
+++ modules/CVS/Entries	(working copy)
@@ -5,7 +5,7 @@
 /FindGLIB.cmake/1.1/Sat Mar  3 14:32:17 2007//
 /FindGNUWIN32.cmake/1.1/Sat Mar  3 14:32:17 2007//
 /FindKDEWIN32.cmake/1.1/Sat Mar  3 14:32:17 2007//
-/FindLibExpat.cmake/1.1/Sat Mar  3 14:32:17 2007//
+/FindLibExpat.cmake/1.2/Fri Mar 16 22:34:40 2007//
 /FindLibIconv.cmake/1.1/Sat Mar  3 14:32:17 2007//
 /FindQt4.cmake/1.1/Sat Mar  3 14:32:17 2007//
 /MacroGetenvWinPath.cmake/1.1/Sat Mar  3 14:32:17 2007//
Index: bus/CVS/Entries
===================================================================
--- bus/CVS/Entries	(revision 610)
+++ bus/CVS/Entries	(working copy)
@@ -1,4 +1,4 @@
-/CMakeLists.txt/1.2/Tue Mar  6 20:51:02 2007//
+/CMakeLists.txt/1.4/Sat Apr 28 19:37:25 2007//
 /service.cmake/1.1/Sat Mar  3 14:32:17 2007//
 /session.conf.cmake/1.1/Sat Mar  3 14:32:17 2007//
 /system.conf.cmake/1.1/Sat Mar  3 14:32:17 2007//
Index: bus/CMakeLists.txt
===================================================================
--- bus/CMakeLists.txt	(revision 610)
+++ bus/CMakeLists.txt	(working copy)
@@ -82,9 +82,9 @@
 include_directories(${XML_INCLUDE_DIR})
 SET (LIBS ${XML_LIBRARY})
 
-add_executable(dbus-daemon ${dbus_daemon_SOURCES})
-target_link_libraries(dbus-daemon dbus-1 ${LIBS})
-install_targets(/bin dbus-daemon )
+add_executable(dbus-daemon${CMAKE_EXE_POSTFIX} ${dbus_daemon_SOURCES})
+target_link_libraries(dbus-daemon${CMAKE_EXE_POSTFIX} dbus-1 ${LIBS})
+install_targets(/bin dbus-daemon${CMAKE_EXE_POSTFIX} )
 install_files(/etc FILES ${config_DATA})
 
 if (DBUS_SERVICE)
@@ -95,9 +95,9 @@
 	#	${BUS_SOURCES} 
 	)
 
-	add_executable(dbus-service ${dbus_service_SOURCES} )
-	target_link_libraries(dbus-service dbus-1 ${LIBS})
-	install_targets(/bin dbus-service )
+	add_executable(dbus-service${CMAKE_EXE_POSTFIX} ${dbus_service_SOURCES} )
+	target_link_libraries(dbus-service${CMAKE_EXE_POSTFIX} dbus-1 ${LIBS})
+	install_targets(/bin dbus-service${CMAKE_EXE_POSTFIX} )
 
 endif (DBUS_SERVICE)
 
@@ -123,7 +123,7 @@
 if (DBUS_BUILD_TESTS)	
 	add_executable(bus-test ${bus_test_SOURCES})
 	target_link_libraries(bus-test dbus-1 ${LIBS} )
-	install_targets(/bin bus-test)
+	#install_targets(/bin bus-test)
 	add_test(bus-test ${EXECUTABLE_OUTPUT_PATH}/bus-test ${CMAKE_SOURCE_DIR}/../test/data)
 endif (DBUS_BUILD_TESTS)
 
Index: CVS/Entries
===================================================================
--- CVS/Entries	(revision 610)
+++ CVS/Entries	(working copy)
@@ -1,4 +1,4 @@
-/CMakeLists.txt/1.4/Wed Mar 14 00:16:05 2007//
+/CMakeLists.txt/1.8/Wed May 23 06:00:34 2007//
 /ConfigureChecks.cmake/1.2/Thu Mar  8 08:36:47 2007//
 /Doxyfile.cmake/1.1/Wed Mar 14 00:16:05 2007//
 /bus-test.bat.cmake/1.1/Sat Mar  3 14:32:17 2007//
Index: CVS/Entries.Log
===================================================================
--- CVS/Entries.Log	(revision 610)
+++ CVS/Entries.Log	(working copy)
@@ -1,5 +1,6 @@
 A D/bus////
 A D/dbus////
+A D/doc////
 A D/modules////
 A D/test////
 A D/tools////
Index: dbus/CVS/Entries
===================================================================
--- dbus/CVS/Entries	(revision 610)
+++ dbus/CVS/Entries	(working copy)
@@ -1,3 +1,3 @@
-/CMakeLists.txt/1.2/Tue Mar  6 20:51:02 2007//
+/CMakeLists.txt/1.7/Thu May 24 19:15:26 2007//
 /dbus-arch-deps.h.cmake/1.1/Sat Mar  3 14:32:17 2007//
 D
Index: dbus/CMakeLists.txt
===================================================================
--- dbus/CMakeLists.txt	(revision 610)
+++ dbus/CMakeLists.txt	(working copy)
@@ -68,6 +68,11 @@
 	    ${DBUS_DIR}/dbus-transport-unix.c
 	    ${DBUS_DIR}/dbus-server-unix.c
 	)
+else(UNIX)
+    set (DBUS_LIB_SOURCES ${DBUS_LIB_SOURCES} 
+#	    ${DBUS_DIR}/dbus-transport-win.c
+	    ${DBUS_DIR}/dbus-server-win.c
+	)
 endif(UNIX)
 
 set (DBUS_LIB_HEADERS
@@ -167,7 +172,7 @@
 if (WIN32)
 	set (DBUS_SHARED_SOURCES ${DBUS_SHARED_SOURCES} 
 		${DBUS_DIR}/dbus-sysdeps-win.c
-		${DBUS_DIR}/dbus-sysdeps-win-thread.c
+		${DBUS_DIR}/dbus-sysdeps-thread-win.c
 		${DBUS_DIR}/dbus-sysdeps-util-win.c
 		${DBUS_DIR}/dbus-spawn-win.c
 	)
@@ -205,7 +210,6 @@
 	${DBUS_UTIL_HEADERS}
 )
 
-set(CMAKE_DEBUG_POSTFIX "d")
 if(MSVC)
 	add_library(dbus-1 STATIC ${libdbus_SOURCES} ${libdbus_HEADERS} )
 	project_source_group(${GROUP_CODE} libdbus_SOURCES libdbus_HEADERS)
@@ -234,10 +238,10 @@
 
 if (DBUS_BUILD_TESTS)
 	set (TESTS_ENVIRONMENT "DBUS_TEST_DATA=${CMAKE_SOURCE_DIR}/test/data DBUS_TEST_HOMEDIR=${CMAKE_BUILD_DIR}/dbus")
-	ADD_EXECUTABLE(dbus-test  ${CMAKE_SOURCE_DIR}/../dbus/dbus-test-main.c)
+	ADD_EXECUTABLE(dbus-test ${CMAKE_SOURCE_DIR}/../dbus/dbus-test-main.c)
 	target_link_libraries(dbus-test dbus-1 ${LIBS})
 	add_test(dbus-test ${EXECUTABLE_OUTPUT_PATH}/dbus-test ${CMAKE_SOURCE_DIR}/../test/data)
-	install_targets(/bin dbus-test)
+	#install_targets(/bin dbus-test)
 ENDIF (DBUS_BUILD_TESTS)
 
 if (UNIX)
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 610)
+++ CMakeLists.txt	(working copy)
@@ -1,6 +1,6 @@
 set (PACKAGE dbus)
 set (VERSION_MAJOR "1")
-set (VERSION_MINOR "0")
+set (VERSION_MINOR "1")
 set (VERSION_PATCH "0")
 set (VERSION ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH} )
 project(${PACKAGE})
@@ -68,14 +68,20 @@
 		set(CMAKE_CXX_WARNING_LEVEL 3 CACHE TYPE STRING FORCE)
 	endif (WALL)
 	
-	SET(MSVC_W_ERROR   " /we4028 /we4013 /we4133 /we4047 /we4031 /we4002 /we4003 /we4114 /we4022")
+	SET(MSVC_W_ERROR   " /we4028 /we4013 /we4133 /we4047 /we4031 /we4002 /we4003 /we4114")
 	SET(MSVC_W_DISABLE " /wd4127 /wd4090 /wd4101 /wd4244")
 
 	SET(CMAKE_C_FLAGS_DEBUG   "${CMAKE_C_FLAGS_DEBUG}   /FIconfig.h ${MSVC_W_ERROR} ${MSVC_W_DISABLE}")
 	SET(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} /FIconfig.h ${MSVC_W_ERROR} ${MSVC_W_DISABLE}")
 endif(MSVC)
 
+if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+   # used by executables, CMAKE_DEBUG_POSTFIX does not handle this case
+   set (CMAKE_EXE_POSTFIX "d")
+   set (CMAKE_DEBUG_POSTFIX "d")
+endif(CMAKE_BUILD_TYPE STREQUAL "Debug")
 
+
 #########################################################################
 # Windows CE
 #
@@ -225,9 +231,6 @@
 #AC_ARG_ENABLE(checks, AS_HELP_STRING([--enable-checks],[include sanity checks on public API]),enable_checks=$enableval,enable_checks=yes)
 OPTION(DBUS_DISABLE_CHECKS "Disable public API sanity checking" OFF)
 
-#AC_ARG_ENABLE(xml-docs, AS_HELP_STRING([--enable-xml-docs],[build XML documentation (requires xmlto)]),enable_xml_docs=$enableval,enable_xml_docs=auto)
-#xmldocs missing
-
 #AC_ARG_ENABLE(gcov, AS_HELP_STRING([--enable-gcov],[compile with coverage profiling instrumentation (gcc only)]),enable_gcov=$enableval,enable_gcov=no)
 OPTION(DBUS_GCOV_ENABLED "compile with coverage profiling instrumentation (gcc only)" OFF)
 if(DBUS_GCOV_ENABLED)
@@ -330,20 +333,12 @@
 	set (EXT ".exe")
 endif(WIN32)
 
-# confirure.in:
-#TEST_PATH(SERVICE_DIR,          data/valid-service-files)
-#TEST_PATH(SERVICE_BINARY,       test-service)
-#TEST_PATH(SHELL_SERVICE_BINARY, test-shell-service)
-#TEST_PATH(EXIT_BINARY,          test-exit)
-#TEST_PATH(SEGFAULT_BINARY,      test-segfault)
-#TEST_PATH(SLEEP_FOREVER_BINARY, test-sleep-forever)
-
 set(TEST_SERVICE_DIR          ${CMAKE_BINARY_DIR}/test/data/valid-service-files     CACHE STRING "Full path to test file test/data/valid-service-files in builddir")
-set(TEST_SERVICE_BINARY       ${EXPANDED_BINDIR}/test-service${EXT}       CACHE STRING "Full path to test file test/test-service in builddir")
-set(TEST_SHELL_SERVICE_BINARY ${EXPANDED_BINDIR}/test-shell-service${EXT} CACHE STRING "Full path to test file test/test-shell-service in builddir")    
-set(TEST_EXIT_BINARY          ${EXPANDED_BINDIR}/test-exit${EXT}          CACHE STRING "Full path to test file test/test-exit in builddir")
-set(TEST_SEGFAULT_BINARY      ${EXPANDED_BINDIR}/test-segfault${EXT}      CACHE STRING "Full path to test file test/test-segfault in builddir")
-set(TEST_SLEEP_FOREVER_BINARY ${EXPANDED_BINDIR}/test-sleep-forever${EXT} CACHE STRING "Full path to test file test/test-sleep-forever in builddir")
+set(TEST_SERVICE_BINARY       ${CMAKE_BINARY_DIR}/bin/test-service${EXT}       CACHE STRING "Full path to test file test/test-service in builddir")
+set(TEST_SHELL_SERVICE_BINARY ${CMAKE_BINARY_DIR}/bin/test-shell-service${EXT} CACHE STRING "Full path to test file test/test-shell-service in builddir")    
+set(TEST_EXIT_BINARY          ${CMAKE_BINARY_DIR}/bin/test-exit${EXT}          CACHE STRING "Full path to test file test/test-exit in builddir")
+set(TEST_SEGFAULT_BINARY      ${CMAKE_BINARY_DIR}/bin/test-segfault${EXT}      CACHE STRING "Full path to test file test/test-segfault in builddir")
+set(TEST_SLEEP_FOREVER_BINARY ${CMAKE_BINARY_DIR}/bin/test-sleep-forever${EXT} CACHE STRING "Full path to test file test/test-sleep-forever in builddir")
 
 #### Find socket directories
 if (NOT WIN32)
@@ -445,8 +440,7 @@
 	add_subdirectory( test )
 endif (DBUS_BUILD_TESTS)
 add_subdirectory( tools )
-# there is currently no directory 'doc'
-#add_subdirectory( doc )
+add_subdirectory( doc )
 
 
 OPTION(DBUS_INSTALL_SYSTEM_LIBS "install required system libraries" OFF)
